### DISCUSSION TITLE
request not aligned to the logical block size

### DISCUSSION THREAD

**Discussion by nguyend24** [openzfs/zfs#15912] [Q&A]
Hello!

This issue came up a couple months ago and I can't figure out any solution online. 
![image](https://github.com/openzfs/zfs/assets/26831516/8136faab-b962-4970-9af2-27536f0193b1)

Every so often under normal operation, this error will show up in dmesg and the disks will get faulted. 

If I do a scrub then it's guaranteed to happen and at minimum 2 of my drives will get faulted (sometimes 4). At first I assumed that the drives had just failed of old age and I replaced 5 of them before getting suspicious. The double drive faults always happen in the same VDEV. It's random drives every time. Sometimes it fails on write errors, sometimes read, seems random.

I replaced the backplane from a Supermicro 846TQ to 846A, but it still occurred.

I then replaced the HBA to an LSI 9207-8, but it still occurs. The HBA is also connected to a Intel RES2SV240.

In the array I have a mix of drives that are 512/4k and 4k native and ashift of 12. Don't know if this information helps but the error seems to imply that it could be an issue.

I have a 24 drive zpool with 8 drives per vdev in RAIDZ2. All drives are 8TB with a mix of various manufacturers and SAS/SATA.

**Comment by rincebrain** [comment]
This could be #15588, maybe, or it could be something stranger; I personally don't recall ever seeing someone report that message. I'm not even sure what it's supposed to mean - the whole point of 512e drives is to pretend to allow 512b aligned accesses, so erroring on that seems strange, and 4kn disks would just reject the request outright.

Are the disks you're getting those from 512e or 4kn?

**Reply by nguyend24** [reply]
![image](https://github.com/openzfs/zfs/assets/26831516/252e40ce-e6d9-4053-92fe-69099a5a523a)

Both drives are 4kn.

**Reply by rincebrain** [reply]
Okay, cute, then yeah, I'd guess for some reason it's issuing requests that aren't 4k-aligned to the 4kn disks and they're unamused.

Guessing #15588, then, at which point my advice would be to try the workaround from [here](https://github.com/openzfs/zfs/issues/15646#issuecomment-1870937865) - that is, shim a 4k sector loopback device between the disk and ZFS, for the moment.

It's gross, but the changes in #15588 are pretty invasive, so we're going to want a lot of testing before putting those in.

**Reply by nguyend24** [reply]
Interesting, the issue you linked mentions the "request not aligned to the logical block size" error that the PR is trying to solve. It didn't come up in any of my searches.

I guess I'll see if I can somewhat replicate this with my data and use the workaround with a smaller test array first.

Thanks!

**Reply by rincebrain** [reply]
I recalled it mostly coming up in the context of things like LUKS or cheap controllers, which would produce errors that are not just "request not aligned". That'll teach me to miss the forest for the trees, even if I got there.

**Reply by nguyend24** [reply]
Putting this here for my reference or anyone else after doing some more searching. I'm trying robn's suggestion of lowering zfs_vdev_aggregation_limit as a workaround.

https://github.com/openzfs/zfs/pull/15414
https://github.com/openzfs/zfs/issues/15533#issuecomment-1825326626
https://github.com/openzfs/zfs/issues/14533#issuecomment-1825329439

**Reply by robn** [reply]
It probably is the same thing described in #15588, though I haven't seen this particular failure mode. I'd love some more info, just to help me confirm:

* OpenZFS version: `zfs version`
* Kernel version: `uname -a`
* What the kernel thinks about those devices: `cat /sys/block/{sdn,sdf}/queue/{logical_block_size,physical_block_size,max_segments,max_segments,max_segment_size,max_sectors_kb,max_hw_sectors_kb}`
* Any other `zio` errors in the kernel log around the same time (within a second or so)
* Any non-standard OpenZFS tuning (kernel parameters, tuneables) around max IO sizes, scrub sizes, scan sizes, etc

Thanks!

**Reply by nguyend24** [reply]
- OpenZFS version: zfs-2.2.2-1
- Kernel: Linux truenas 6.1.63-production+truenas 2 SMP PREEMPT_DYNAMIC Tue Jan 23 22:31:02 UTC 2024 x86_64 GNU/Linux
- cat /sys/block/{sdn,sdf}/queue/{logical_block_size,physical_block_size,max_segments,max_segments,max_segment_size,max_sectors_kb,max_hw_sectors_kb}
4096
4096
128
128
65536
1280
16380
4096
4096
128
128
65536
1280
16380
- No other zio errors
- Everything should be default unless Truenas Scale changed it

**Reply by robn** [reply]
Couple more questions: what's your `recordsize` (all of them, if mixed) and what's your pool capacity% and fragmentation% at?

**Reply by nguyend24** [reply]
Record size on all datasets is the default 128K

![image](https://github.com/openzfs/zfs/assets/26831516/5d032a83-a769-4089-85af-383fa08a5fe1)

**Reply by ossimoi** [reply]
Are the faulting drives always specific type or mixture of SAS and SATA?

**Reply by nguyend24** [reply]
I don't remember. I didn't check that. 

Happy to report though that after lowering zfs_vdev_aggregation_limit to 500000, I'm finally able to complete a scrub without any drives faulting or checksum errors

**Reply by robn** [reply]
Well, glad to see that. Little uncertain as to why; I didn't think scrub reads were aggregated through the vdev queue, so I didn't think that setting woud make a difference in this case. Clearly, I was wrong.

Its definitely #15588 though. Working on it!

**Reply by HarryFreeMyLand** [reply]
I'd also like to report after having ran into this issue myself, deceasing the parameter as suggested above has helped my pool avoid this error and failed I/O (I've not yet seen it re-occur, even during scrub+copying).

in my case, it's a 8x 8-wide RAIDz2, all 4kn, scrubs and other high I/O would result in these errors and faulted devices. host system is a TrueNAS Scale system running 24.04, running openzfs 2.2.3.

but I imagine that any system with a number of 4kn disks will run into this issue.
